### ISSUE TITLE
Services with long-lived connections are not load balanced

### ISSUE THREAD

**Issue by ammills01** [microsoft/presidio#304]
**Describe the bug**
When API instances spool up they create long-lived connections to other services, like the Analyzer. Even though that connection is created using the Cluster IP Service URL the API locks onto a single pod and the load is not balanced.

**To Reproduce**
1. Stand up API with 1 pod
2. Stand up Analyzer with 2+ pods
3. Open logs for all Analyzer pods and set auto-refresh
4. Make 10+ Analyzer/Anonymize calls to the API
5. Notice only 1 Analyzer pod sees traffic

**Expected behavior**
Load should have been evenly balanced among the pods behind the Cluster IP Service.

**Additional context**
This article does a really good job to explain it:  https://learnk8s.io/kubernetes-long-lived-connections

Basically, Kubernetes doesn't load balance long-lived connections. This applies to connections using HTTP/2, gRPC, RSockets, AMQP, etc...

I was able to fix this by deploying Istio to my cluster and setting up injections on the presidio namespace.

**Comment by balteravishay** [comment]
great catch! 
we are considering linkerd/istio as well.

**Comment by ammills01** [comment]
I have not messed with Linkerd, but installing Istio on the AKS was super easy. I followed the guide on the MS website and it probably took me 10 minutes. The Kiali UI is really intuitive and easy to use as well. 

https://docs.microsoft.com/en-us/azure/aks/servicemesh-istio-install?pivots=client-operating-system-linux

Personally, I'd like to see Istio or Linkerd as an option when setting up an AKS cluster. Check a box, supply credentials, and your cluster stands up with a mesh already configured. However, a service mesh may not be for everyone. Have you all considered changing the project to use something like [Ribbon](https://github.com/Netflix/ribbon )? That would allow it to work out of the box instead of requiring the user install some sort of mesh.

No matter which way you go I think we need to get something in the documentation that clearly points out the issue and mentions solutions. I've tweaked the Helm script to take a parameter for Analyzer replicas, that might be a good spot to put something.

**Comment by LizaShak** [comment]
Thank you for your feedback @ammills01 , we have added the documentation task to our backlog.

**Comment by ammills01** [comment]
Just wanted to update this thread to say that we moved away from Istio and started using Linkerd for our service mesh. Istio got out of hand pretty quick and made our cluster unusable due to the number of resources it was consuming. So far, Linkerd has not had that same issue and was just as easy to deploy and inject into the Presidio pods.

**Comment by diwu1989** [comment]
Isn't this easier to solve by not re-using client and long-lived connections?
Is the setup cost of a RPC client that expensive compared to the analysis of the payload?

**Comment by ammills01** [comment]
> Isn't this easier to solve by not re-using client and long-lived connections?
> Is the setup cost of a RPC client that expensive compared to the analysis of the payload?

That depends on your volume. I'm churning 800+ records per second through Presidio in production so the loss in efficiency would be pretty painful. Also, Linkerd was really easy to implement.

**Comment by maitham** [comment]
@ammills01  are you able to share specs of cluster etc. We seem to be limited to approx 600 requests per second regardless of how we scale pods, we're using linkerd too.

**Comment by ammills01** [comment]
Sure @maitham 

Azure Kubernetes Services running 5 Standard_F8s VS's in EastUS 2. I've got an HPA on but can generally achieve 800+ rps with 10-12 Analyzer pods, and I scale Ingress and API pods to be 1/4 the number of Analyze pods. Resource limits for Analyzer pods are set at 1 CPU and 5 GB of RAM. I've found passing in a template works much faster than passing  a reference  and having Presidio hit a Redis cache, though the test I did was Redis in Azure but not in k8s.

I'm streaming data from Kafka using KStreams and making API calls to Presidio. I've got 24 partitions on the Kafka topic so I can achieve 24 simultaneous calls, and I run 24 threads of the KStreams job, across 3 VM's (not containerized yet), to get max output. I've entertained the idea of setting up a perpetual job in Presidio to consume from the Kafka topic but it's not designed for that and there are other considerations that make the API the best solution for us.

As I write this I realize my aforementioned RPS measurements come from the next hop in the pipeline, a Spark job. The data I'm working with may or may not need redacted, so some of the records don't actually call Presidio. That's not a significant number, certainly less than 50 reps, but worth noting.

**Comment by navalev** [comment]
@ammills01 We recently released the new version on presidio V2, whcih includes updates to the deployment samples https://github.com/microsoft/presidio/tree/main/docs/samples/deployments 
This issue will be closed. Feel free to re-open in case this is not resolved.
